src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -770,6 +770,77 @@ internal override bool IsExplicitDefinitionOfNoPiaLocalType
             }
         }
 
+        private void CheckPresenceOfTypeIdentifierAttribute()
+        {
+            // We want this function to be as cheap as possible, it is called for every top level type
+            // and we don't want to bind attributes attached to the declaration unless there is a chance
+            // that one of them is TypeIdentifier attribute.
+            ImmutableArray<SyntaxList<AttributeListSyntax>> attributeLists = GetAttributeDeclarations();
+
+            foreach (SyntaxList<AttributeListSyntax> list in attributeLists)
+            {
+                var syntaxTree = list.Node.SyntaxTree;
+                Binder binder = null;
+
+                foreach (AttributeListSyntax attrList in list)
+                {
+                    foreach (AttributeSyntax attr in attrList.Attributes)
+                    {
+                        SimpleNameSyntax simpleName = attr.Name.GetUnqualifiedName();
+                        string name = simpleName.Identifier.ValueText;
+
+                        if (name == "TypeIdentifier" || name == AttributeDescription.TypeIdentifierAttribute.Name)
+                        {
+                            // This attribute syntax might be an application of TypeIdentifierAttribute.
+                            // Let's bind it.
+                            // For simplicity we bind all attributes.
+                            GetAttributes();
+                            return;
+                        }
+
+                        if (attr.Name == simpleName)
+                        {
+                            // Check if the name matches an alias bound to the TypeIdentifierAttribute
+                            if (binder == null)
+                            {
+                                binder = this.DeclaringCompilation.GetBinderFactory(list.Node.SyntaxTree).GetBinder(list.Node);
+                            }
+
+                            string nameWithAttribute = name + "Attribute";
```

Comment:
❓ Can we lazily initialize this after we know `usingAliases` is not empty?

Reply:
> Can we lazily initialize this after we know usingAliases is not empty?

Sure, we can do this.